core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -315,7 +319,7 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorAdded(executorId: String): Unit = synchronized {
     if (!executorIds.contains(executorId)) {
       executorIds.add(executorId)
-      executorIds.foreach(onExecutorIdle)
+      executorIds.filter(listener.isExecutorIdle).foreach(onExecutorIdle)
```

Comment:
This deserves a loud comment, perhaps something like:
```
If an executor is not removed because the lower bound has been reached, mark it idle again when new executors join because we are no longer at the lower bound. Otherwise, we may never remove these executors until after they run tasks again.
```